nautobot/ipam/utils/migrations.py

```diff
@@ -0,0 +1,569 @@
+    check_interface_vrfs(apps)
+
+    # VRFs
+    process_vrfs(apps)
+
+    add_prefix_and_ip_address_version(apps)
+
+    # IPAddresses
+    # TODO: deduplicate IP addresses
+    process_ip_addresses(apps)
+
+    # Prefixes
+    process_prefix_duplicates(apps)
+    reparent_prefixes(apps)
+
+    # Make another pass across all VRFs to duplicate it if it has prefixes
+    # in another namespace (non-unique VRFs with duplicate Prefixes)
+    copy_vrfs_to_cleanup_namespaces(apps)
+
+    # [VM]Interfaces
+    process_interfaces(apps)
+
+    # VRF-Prefix M2M
+    process_vrfs_prefixes_m2m(apps)
+
+
+def check_interface_vrfs(apps):
+    """
+    Enumerate all Interface and VMInterface objects and raise an exception if any interface is found that is associated
+    to more than one distinct VRF through the ip_address many-to-many relationship.
+    """
+
+    Interface = apps.get_model("dcim", "Interface")
+    VMInterface = apps.get_model("virtualization", "VMInterface")
+
+    interfaces_with_multiple_vrfs = Interface.objects.annotate(vrf_count=models.Count("ip_addresses__vrf")).filter(
+        vrf_count__gt=1
+    )
+    vm_interfaces_with_multiple_vrfs = VMInterface.objects.annotate(vrf_count=models.Count("ip_addresses__vrf")).filter(
+        vrf_count__gt=1
+    )
+
+    if interfaces_with_multiple_vrfs.exists() or vm_interfaces_with_multiple_vrfs.exists():
+        raise Exception(
+            "You cannot migrate Interfaces or VMInterfaces that have IPs with differing VRFs.",
+            list(interfaces_with_multiple_vrfs),
+            list(vm_interfaces_with_multiple_vrfs),
+        )
+
+
+def process_vrfs(apps):
+    """
+    Enumerate all VRF objects in the database and attempt to find suitable Namespace with no duplicate VRF names and
+    no duplicate Prefixes associated to the VRF. Any VRF with `enforce_unique` set and has related prefixes will be
+    moved to its own Namespace. All other VRFs will be checked for duplicate names and Prefixes and moved to a cleanup
+    Namespace if any duplicates are found.
+
+    Args:
+        apps: Django apps module
+
+    Returns:
+        None
+    """
+    VRF = apps.get_model("ipam", "VRF")
+    vrfs = VRF.objects.all().order_by("name", "rd")
+    unique_non_empty_vrfs = vrfs.filter(enforce_unique=True).exclude(ip_addresses__isnull=True, prefixes__isnull=True)
+    global_ns_vrfs = vrfs.filter(namespace__name="Global")
+
+    # Case 0: VRFs with enforce_unique move to their own Namespace.
+    # TODO(jathan): Should we also check for Prefix overlap here, too?
+    for vrf in unique_non_empty_vrfs:
+        print(f">>> Processing migration for VRF {vrf.name!r}, Namespace {vrf.namespace.name!r}")
+        vrf.namespace = create_vrf_namespace(apps, vrf)
+        vrf.save()
+        vrf.prefixes.update(namespace=vrf.namespace)
+        print(f"    VRF {vrf.name!r} migrated to Namespace {vrf.namespace.name!r}")
+
+    # Case 00: Unique fields vs. Namespaces (name/rd) move to a Cleanup Namespace.
```

Comment:
Can this be reworded? I'm not clear what `unique fields vs. Namespaces` means here.